_posts/2019-01-10-future-workloads.md

```diff
@@ -0,0 +1,503 @@
+Thread[scala-execution-context-global-11,5,main]
+Thread[scala-execution-context-global-14,5,main]
+Thread[scala-execution-context-global-17,5,main]
+Thread[scala-execution-context-global-13,5,main]
+... continues until 1000 executions
+```
+
+`htop` will now print all cores at 100% utilization. The OS will manage the extra threads and put them to sleep, leaving our cpu threads mostly unaffected.
+
+The downside is that spawning more threads than there are cores will eventually degrade performance on cpu bound tasks. Also, we are increasing the amount of context switches and memory usage per new thread. Finally, since system calls are expensive, this is also hurting our performance.
+
+The only way around blocking is to go asynchronous, which we will look at next.
+
+## Asynchronous
+
+Asynchronous calls are fast because a callback informs us of a result, so we are not wasting time waiting. We can understand this by looking at our `async` method. The scheduler is simulating a client side.
+
+Let's take for example an arbitrary request. While blocking request may look like this:
+```
+time 1: Request value x
+time 2: wait...
+time 3: wait...
+time 4: Obtain x
+time 5: Request value y
+time 6: wait...
+time 7: wait...
+time 8: Obtain y
+```
+
+Async will be more like this:
+```
+time 1: Request value x
+time 2: Request value y
+time 3:
+time 4: Obtain x
+time 5: Obtain y
+```
+
+What is the empty space in the async example? Any other work the cpu could be doing. Asynchronous code does not occupy time gaps with waiting.
+
+*Note: The OS will put blocking threads to sleep, so the first example above is actually more efficient than shown here.*
+
+Asynchronous futures are of a different species. All 1000 futures will be complete in ~1000ms.
+
+```
+Future.sequence(List.fill(1000)(async(1000))).await()
+```
+
+```
+... all previous executions
+Thread[scala-execution-context-global-14,5,main]
+Thread[scala-execution-context-global-14,5,main]
+Thread[scala-execution-context-global-14,5,main]
+Thread[scala-execution-context-global-14,5,main]
+Thread[scala-execution-context-global-14,5,main]
+Thread[scala-execution-context-global-13,5,main]
+Thread[scala-execution-context-global-12,5,main]
+End
+```
+
+These being so fast means there is less of a problem with scheduling them in an heterogeneous workload containing `cpuBound` calls, for example.
+
+Also note that we have to be careful managing asynchronous futures, as we can go out of memory by creating large amounts of Future objects.
+
+#### Thread Pools
+
+So far we have used the default `scala.concurrent thread pool` and a `fixed thread pool` in our simple examples. We have scheduled cpu heavy futures in the first, and blocking futures in the second, in order to achieve more throughput.
+
+The Java [Executors](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Executors.html) class can spawn other executors, and we can convert them to an `ExecutionContext` as such:
```

Comment:
let java 7 rest in peace,
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/Executors.html

Reply:
it shall rip